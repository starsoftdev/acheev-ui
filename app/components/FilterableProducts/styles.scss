@import '~styles/settings/settings';

.filterableProducts {
  @include breakpoint(small only) {
    margin-bottom: $rem-md;
  }
  &__item {
    padding-top: $rem-xl;
    padding-bottom: $rem-xl;
    margin-bottom: $rem-xl;
    @include breakpoint(small only) {
      padding-top: $rem-tn;
      padding-bottom: $rem-sm;
      margin-bottom: $rem-sm;
      border-bottom: none;
    }
  }
  &__titleContainer {
    @include breakpoint(small only) {
      margin-bottom: $rem-sm;
    }
  }
  &__title {
    text-transform: capitalize;
    margin: 0;
    line-height: 1;
    @include breakpoint(small only) {
      margin: $rem-sm 0;
    }
  }
  &__showAll {
    @include breakpoint(small only) {
      float: right;
    }
  }
  &__buttonGroup {
    display: inline-flex;
  }
  &__buttonGroupItem {
    width: rem-calc(140);
  }
  &__carousel {
    > div { // there's no possibility to set a className for unslicked carousel
      @include breakpoint(medium) {
        display: flex;
        flex-flow: row wrap;
      }
    }
    .product {
      @include breakpoint(small only) {
        padding-bottom: $rem-xl;
      }
    }
    .bb-light-gray {
      @include breakpoint(small only) {
        border-bottom: none !important;
      }
    }
  }
}
