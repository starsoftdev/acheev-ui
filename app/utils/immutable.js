// @flow

import { Iterable } from 'immutable';
import type { Map } from 'immutable';
import { get as _get } from 'lodash-es';

export const get = (
  collection: ?Map<*, *> | Object,
  path: string | number,
  defaultValue: any
) => {
  if (!collection) {
    return defaultValue;
  } else if (Iterable.isIterable(collection)) {
    return collection.get(path, defaultValue);
  }
  return _get(collection, path, defaultValue);
};

export const getIn = (
  collection: ?Map<*, *> | Object,
  path: Array<string | number>,
  defaultValue: any
) => {
  if (!collection) {
    return defaultValue;
  } else if (Iterable.isIterable(collection)) {
    return collection.getIn(path, defaultValue);
  }
  return _get(collection, path, defaultValue);
};
