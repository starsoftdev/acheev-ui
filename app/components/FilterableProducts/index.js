// @flow

import * as React from 'react';
import Link from 'components/Link';
import cx from 'classnames';
import type { Map } from 'immutable';

import Carousel from 'components/Carousel';
import Preloader from 'components/Preloader';
import ButtonGroup, { ButtonGroupItem } from 'components/ButtonGroup';
import ProductCard from 'components/ProductCard';
import Select from 'components/Select';
import './styles.scss';

type Props = {
  data: Map<*, *>,
  category: string,
  requestItems: Function,
  linksTo: string,
  border: ?Boolean,
};

const carouselSettings = {
  responsive: [
    {
      breakpoint: 640,
      settings: {
        dots: true,
        arrows: false,
      },
    },
    {
      breakpoint: 9999,
      settings: 'unslick',
    },
  ],
};

const FilterableProducts = ({
  data,
  category,
  requestItems,
  linksTo,
  border,
}: Props) => {
  if (!data) return null;
  const items = data.getIn(['filters', data.get('active'), 'items']);
  const filters = data.get('filters');
  return (
    <div className="row column">
      <div className="filterableProducts">
        <div className="row align-bottom filterableProducts__titleContainer">
          <div className="small-12 medium-3 large-4 column">
            <h2 className="filterableProducts__title">
              {category}
              <span className="filterableProducts__showAll show-for-small-only">
                <Link className="fs-md" to={linksTo}>
                  See all
                </Link>
              </span>
            </h2>
          </div>
          <div className="small-12 medium-7 large-6 column">
            <div className="hide-for-small-only">
              <ButtonGroup className="filterableProducts__buttonGroup">
                {filters.map(item => {
                  const index = item.get('index');
                  return (
                    <ButtonGroupItem
                      active={item.get('index') === data.get('active')}
                      key={index}
                      onClick={() => requestItems(index, category)}
                      className="filterableProducts__buttonGroupItem"
                    >
                      {item.get('name')}
                    </ButtonGroupItem>
                  );
                })}
              </ButtonGroup>
            </div>
            <div className="show-for-small-only">
              <Select
                className="b-primary"
                value={data.get('active')}
                onChange={e => requestItems(e.target.value, category)}
              >
                {filters.map(item => {
                  const index = item.get('index');
                  return (
                    <option key={index} value={index}>
                      {item.get('name')}
                    </option>
                  );
                })}
              </Select>
            </div>
          </div>
          <div className="small-12 medium-2 large-2 column text-right hide-for-small-only">
            <Link className="fs-md" to={linksTo}>
              See all
            </Link>
          </div>
        </div>
        {data.get('isLoading') ? (
          <Preloader height={423} />
        ) : (
          <div className="filterableProducts__carousel">
            <Carousel {...carouselSettings}>
              {items &&
                items.map(item => (
                  <div
                    className={cx(
                      'filterableProducts__item small-12 medium-6 large-3 column',
                      { 'bb-light-gray': border }
                    )}
                    key={item.get('_id')}
                  >
                    <ProductCard data={item} alignCenter />
                  </div>
                ))}
            </Carousel>
          </div>
        )}
      </div>
    </div>
  );
};

export default FilterableProducts;
