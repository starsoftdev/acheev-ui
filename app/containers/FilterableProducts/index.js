// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';

import FilterableProducts from 'components/FilterableProducts';
import { requestFilterableProducts } from 'pages/Home/sagas';

type Props = {
  data: Object,
  category: string,
  requestFilterableProducts: Function,
  linksTo: string,
  border: ?Boolean,
};

class FilterableProductsContainer extends Component<Props> {
  componentWillMount() {
    this.props.requestFilterableProducts(0, this.props.category);
  }
  render() {
    const { data, linksTo, border } = this.props;
    return (
      <FilterableProducts
        category={this.props.category}
        data={data}
        linksTo={linksTo}
        requestItems={(index, category) =>
          this.props.requestFilterableProducts(index, category)}
        border={border}
      />
    );
  }
}

const mapStateToPtops = (state, props) => ({
  data: state.getIn(['home', props.category]),
});

const mapDispatchToProps = dispatch => ({
  requestFilterableProducts: (index, category) =>
    dispatch(requestFilterableProducts(index, category)),
});

export default connect(mapStateToPtops, mapDispatchToProps)(
  FilterableProductsContainer
);
