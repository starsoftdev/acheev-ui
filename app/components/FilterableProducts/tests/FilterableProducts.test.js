import React from 'react';
import { shallow, mount } from 'enzyme';
import { fromJS } from 'immutable';
import { create } from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

import FilterableProducts from 'components/FilterableProducts';

const data = fromJS({
  url: 'Strain',
  filters: [
    {
      index: 0,
      name: 'Most popular',
      sort: '-reviewCount',
      items: [
        {
          business: {
            photos: [],
            hoursOfOperation: [],
            telephone: '1-855-55-TWEED(89333)',
            admins: [6267],
            facebook: 'https://www.facebook.com/tweedinc',
            createdOn: '2014-12-11T05:06:00.238Z',
            productReviewCount: 7652,
            name: 'Tweed',
            slug: 'tweed',
            thumbnail: 'https://lift.co/uploads/v6AIjoUMF7ozJw4UWzgF.png',
            productRating: 4,
            published: true,
            registrationLink: 'http://www.tweed.com/account/register',
            twitter: 'https://twitter.com/tweedinc',
            __t: 'Producer',
            __v: 0,
            reviewCount: 2599,
            tags: [],
            _id: '54892638f2c99094235ef273',
            updatedOn: '2017-08-01T17:19:24.354Z',
            reviews: ['5980b8660952e830dce85e20', '5980b8660952e830dce85e0e'],
            website: 'http://tweed.com',
            id: '54892638f2c99094235ef273',
            email: 'hi@tweed.com',
            description:
              '<p>Tweed is a Canadian company established to supply an unmatched selection of premium medical marijuana to treat symptoms such as chronic pain, seizures, muscle spasms, nausea and loss of appetite. We believe in providing only reliable, high quality products, and strive to do so with the utmost empathy, compassion, professionalism and integrity.</p>',
            features: {
              promoted: false,
              points: true,
            },
            rating: 4,
            fax: '1-888-977-2595',
            linkedin: 'https://www.linkedin.com/company/tweed-inc-',
            locations: [
              {
                province: 'Ontario',
                city: 'Smiths Falls',
                address: '1 Hershey Dr',
                postalCode: 'K7A 0A8',
                _id: '5980b81c0952e830dce6152c',
                latLon: [44.8997225, -76.00185320000003],
              },
            ],
          },
          photos: [],
          flavours: [],
          thcLow: 8.04,
          cbd: '5.93%',
          createdOn: '2015-05-13T21:42:12.154Z',
          cbdHigh: 5.93,
          positiveEffects: [],
          symptoms: [],
          name: 'Penelope',
          slug: 'tweed-penelope',
          negativeEffects: [],
          variants: [
            {
              price: 900,
              doseUnit: 'g',
              doseAmount: 1,
              _id: '5980b81e0952e830dce618fb',
              weightUnit: 'g',
              availabilityStatus: 'available',
              available: true,
            },
          ],
          published: true,
          thc: '8.04%',
          cbdLow: 5.93,
          thcHigh: 8.04,
          searchField: 'Penelope Tweed hybrid',
          prescribedFor: [],
          __t: 'Strain',
          awards: [],
          reviewCount: 542,
          title: 'Penelope',
          tags: [],
          _id: '5980b81e0952e830dce618fa',
          updatedOn: '2017-05-25T23:25:32.874Z',
          reviews: ['5980b8200952e830dce61dbe', '5980b8200952e830dce61ec0'],
          id: '5980b81e0952e830dce618fa',
          milled: false,
          description: '<p><span>CBD Skunk Haze</span></p><p><br/></p>',
          rating: 3.5,
          category: 'hybrid',
        },
      ],
    },
    {
      index: 1,
      name: 'Trending',
      sort: '-rating,-createdOn',
      items: null,
    },
    {
      index: 2,
      name: 'New',
      sort: '-createdOn',
      items: null,
    },
  ],
  isLoading: false,
  active: 0,
});

jest.mock('react-dotdotdot', () => 'div');
jest.mock('react-rating', () => 'div');
jest.mock('react-slick', () => 'div');
// eslint-disable-next-line no-global-assign
console = {
  log: jest.fn(),
  error: jest.fn(),
};

describe('<FilterableProducts />', () => {
  it("shouldn't throw error if data is not specified", () => {
    try {
      shallow(
        <MemoryRouter>
          <FilterableProducts />
        </MemoryRouter>
      );
    } catch (error) {
      expect(error).toBeUndefined();
    }
  });

  it('addToCart should be fired when Button is clicked', () => {
    const spy = jest.fn();
    const renderedComponent = mount(
      <MemoryRouter>
        <FilterableProducts requestItems={spy} data={data} />
      </MemoryRouter>
    );
    renderedComponent
      .find('ButtonGroupItem')
      .first()
      .simulate('click');
    expect(spy).toHaveBeenCalled();
  });

  it('should render Link with appropriate `to` prop when `linksTo` prop is specified', () => {
    const renderedComponent = mount(
      <MemoryRouter>
        <FilterableProducts data={data} linksTo="/hey" />
      </MemoryRouter>
    );
    expect(
      renderedComponent
        .find('Link')
        .at(0)
        .prop('to')
    ).toEqual('/hey');
    expect(
      renderedComponent
        .find('Link')
        .at(1)
        .prop('to')
    ).toEqual('/hey');
  });

  it('should render category if category prop is specified', () => {
    const renderedComponent = mount(
      <MemoryRouter>
        <FilterableProducts data={data} category="Hey" />
      </MemoryRouter>
    );
    expect(renderedComponent.text()).toContain('Hey');
  });

  it('should add className `bb-light-gray` to `.filterableProducts__item` if `border` prop is specified', () => {
    const renderedComponent = mount(
      <MemoryRouter>
        <FilterableProducts data={data} border />
      </MemoryRouter>
    );
    expect(
      renderedComponent
        .find('.filterableProducts__item')
        .hasClass('bb-light-gray')
    ).toEqual(true);
  });

  it('renders data if provided', () => {
    const component = create(
      <MemoryRouter>
        <FilterableProducts data={data} />
      </MemoryRouter>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
