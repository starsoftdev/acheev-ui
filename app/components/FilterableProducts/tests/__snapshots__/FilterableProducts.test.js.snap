// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterableProducts /> renders data if provided 1`] = `
<div
  className="row column"
>
  <div
    className="filterableProducts"
  >
    <div
      className="row align-bottom filterableProducts__titleContainer"
    >
      <div
        className="small-12 medium-3 large-4 column"
      >
        <h2
          className="filterableProducts__title"
        >
          <span
            className="filterableProducts__showAll show-for-small-only"
          >
            <a
              className="fs-md"
            >
              See all
            </a>
          </span>
        </h2>
      </div>
      <div
        className="small-12 medium-7 large-6 column"
      >
        <div
          className="hide-for-small-only"
        >
          <div
            className="button-group filterableProducts__buttonGroup"
          >
            <button
              className="button active filterableProducts__buttonGroupItem"
              onClick={[Function]}
            >
              <div
                className="row align-middle align-center"
              >
                <div
                  className="shrink column"
                >
                  Most popular
                </div>
              </div>
            </button>
            <button
              className="button light hollow filterableProducts__buttonGroupItem"
              onClick={[Function]}
            >
              <div
                className="row align-middle align-center"
              >
                <div
                  className="shrink column"
                >
                  Trending
                </div>
              </div>
            </button>
            <button
              className="button light hollow filterableProducts__buttonGroupItem"
              onClick={[Function]}
            >
              <div
                className="row align-middle align-center"
              >
                <div
                  className="shrink column"
                >
                  New
                </div>
              </div>
            </button>
          </div>
        </div>
        <div
          className="show-for-small-only"
        >
          <select
            className="b-primary"
            onChange={[Function]}
            value={0}
          >
            <option
              value={0}
            >
              Most popular
            </option>
            <option
              value={1}
            >
              Trending
            </option>
            <option
              value={2}
            >
              New
            </option>
          </select>
        </div>
      </div>
      <div
        className="small-12 medium-2 large-2 column text-right hide-for-small-only"
      >
        <a
          className="fs-md"
        >
          See all
        </a>
      </div>
    </div>
    <div
      className="filterableProducts__carousel"
    >
      <div
        className=""
        nextArrow={<SampleNextArrow />}
        prevArrow={<SamplePrevArrow />}
        responsive={
          Array [
            Object {
              "breakpoint": 640,
              "settings": Object {
                "arrows": false,
                "dots": true,
              },
            },
            Object {
              "breakpoint": 9999,
              "settings": "unslick",
            },
          ]
        }
        shouldComponentUpdate={true}
      >
        <div
          className="filterableProducts__item small-12 medium-6 large-3 column"
        >
          <div
            className="productCard"
          >
            <a
              className="productCard__link"
              href="/strains/tweed-penelope"
              onClick={[Function]}
            >
              <div
                aria-haspopup={true}
                className="productCard__image"
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "backgroundImage": "url('https://images.lift.co/resize/78x78/https://lift.co/uploads/v6AIjoUMF7ozJw4UWzgF.png')",
                  }
                }
              >
                <svg
                  className="productCard__categoryIcon"
                  height={30}
                  width={30}
                >
                  <use
                    xlinkHref="#undefined"
                  />
                </svg>
              </div>
              <h4
                className="productCard__title"
              >
                <div
                  clamp={2}
                >
                  Penelope
                </div>
              </h4>
            </a>
            <a
              className="productCard__producer"
              href="/producers/tweed"
              onClick={[Function]}
            >
              Tweed
            </a>
            <div
              className="cannabinoids row align-center"
            >
              <div
                className="shrink column t-nowrap"
              >
                <strong>
                  THC
                </strong>
                 
                8%
              </div>
              <div
                className="shrink column t-nowrap"
              >
                <strong>
                  CBD
                </strong>
                 
                5.9%
              </div>
            </div>
            <div
              className="reviewCount row align-middle align-center mb-mn"
            >
              <div
                className="reviewCount__starSection column small-12 medium-shrink"
              >
                <div
                  className="reviewCount__starsContainer row align-middle align-center"
                >
                  <div
                    className="shrink column"
                  >
                    <div
                      className="starRating reviewCount__star"
                    >
                      <div
                        emptySymbol={
                          <Icon
                            glyph="IMAGE_MOCK"
                            size={16}
                          />
                        }
                        fractions={2}
                        fullSymbol={
                          <Icon
                            glyph="IMAGE_MOCK"
                            size={16}
                          />
                        }
                        initialRating={3.5}
                        readonly={true}
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                className="reviewCount__countSection column small-12 medium-shrink"
              >
                <strong>
                  (
                  3.5
                  )   
                </strong>
                <a
                  className=""
                  href="/strains/tweed-penelope/reviews"
                  onClick={[Function]}
                >
                  2
                   
                  
                  reviews
                </a>
              </div>
            </div>
            <div
              className="row"
            >
              <div
                className="column text-center"
              >
                <a
                  className="button secondary small"
                  onClick={[Function]}
                >
                  <div
                    className="row align-middle align-center"
                  >
                    <div
                      className="shrink column"
                    >
                      View
                    </div>
                  </div>
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
